ms2deepscore/train_new_model/train_ms2deepscore.py

```diff
@@ -3,44 +3,65 @@
 from ms2deepscore import SpectrumBinner
-from ms2deepscore.data_generators import DataGeneratorAllInchikeys
+from ms2deepscore.data_generators import DataGeneratorAllInchikeys, DataGeneratorAllSpectrums
 from ms2deepscore.models import SiameseModel
 from ms2deepscore.train_new_model.calculate_tanimoto_matrix import calculate_tanimoto_scores_unique_inchikey
+from ms2deepscore.utils import save_pickled_file, return_non_existing_file_name, load_pickled_file
 
 
-def train_ms2ds_model(training_spectra,
-                      validation_spectra,
-                      tanimoto_df,
-                      output_model_file_name,
-                      epochs=150):
-    assert not os.path.isfile(output_model_file_name), "The MS2Deepscore output model file name already exists"
+def bin_spectra(training_spectra: List[Spectrum],
+                validation_spectra: List[Spectrum],
+                additional_metadata=(),
+                save_folder=None):
     # Bin training spectra
     spectrum_binner = SpectrumBinner(10000, mz_min=10.0, mz_max=1000.0, peak_scaling=0.5,
-                                     allowed_missing_percentage=100.0)
+                                     allowed_missing_percentage=100.0,
+                                     additional_metadata=additional_metadata)
     binned_spectrums_training = spectrum_binner.fit_transform(training_spectra)
     # Bin validation spectra using the binner based on the training spectra.
     # Peaks that do not occur in the training spectra will not be binned in the validaiton spectra.
     binned_spectrums_val = spectrum_binner.transform(validation_spectra)
+    if save_folder:
+        if not os.path.exists(save_folder):
+            assert not os.path.isfile(save_folder), "The folder specified is a file"
+            os.mkdir(save_folder)
+        save_pickled_file(binned_spectrums_training, return_non_existing_file_name(os.path.join(save_folder, "binned_training_spectra.pickle")))
+        save_pickled_file(binned_spectrums_val, return_non_existing_file_name(os.path.join(save_folder, "binned_validation_spectra.pickle")))
+        save_pickled_file(spectrum_binner, return_non_existing_file_name(os.path.join(save_folder, "spectrum_binner.pickle")))
+    return binned_spectrums_training, binned_spectrums_val, spectrum_binner
+
+
+def train_ms2ds_model(binned_spectrums_training,
+                      binned_spectrums_val,
+                      spectrum_binner,
+                      tanimoto_df,
+                      output_model_file_name,
+                      epochs=150,
+                      base_dims=(500, 500),
+                      embedding_dim=200
+                      ):
+    assert not os.path.isfile(output_model_file_name), "The MS2Deepscore output model file name already exists"
 
     same_prob_bins = list(zip(np.linspace(0, 0.9, 10), np.linspace(0.1, 1, 10)))
 
-    training_generator = DataGeneratorAllInchikeys(
+    training_generator = DataGeneratorAllSpectrums(
```

Comment:
I would still go for `DataGeneratorAllInchikeys`. I looked at the code again and there is indeed something confusing about it. In future PRs I would probably skip the parameter `selected_inchikeys` which only makes things more complicated. Still, the "AllSpectrums" will go through al spectrums which not only makes training longer, it will also shift the bias towards very frequently occurring compounds.